go-controller/pkg/ovn/controller/egressservice_zone/egressservice_zone.go

```diff
@@ -594,114 +503,85 @@ func (c *Controller) syncEgressService(key string) error {
 	}
 
 	state := c.services[key]
-	if es == nil {
+
+	// Clean up the service if it is not assigned to any host or was removed
+	if es == nil || len(es.Status.Host) == 0 {
+		klog.V(5).Infof("Egress service %s was removed or is not assigned to any host", key)
 		if state == nil {
-			// The EgressService was deleted and wasn't configured before.
-			// We delete it from the unallocated service cache just in case.
-			delete(c.unallocatedServices, key)
+			// The egress service was not configured, nothing to do
 			return nil
 		}
-		// The service is configured but does no longer have the egress service resource,
+		// The egress service is configured, but it no longer exists/is assigned to a node,
 		// meaning we should clear all of its resources.
 		return c.clearServiceResourcesAndRequeue(key, state)
 	}
 
 	if svc == nil {
+		klog.V(5).Infof("Service %s doesn't exist", key)
 		if state == nil {
-			// The service object was deleted and was not an allocated egress service.
-			// We delete it from the unallocated service cache just in case.
-			delete(c.unallocatedServices, key)
-			return c.setEgressServiceHost(namespace, name, "")
+			// The service object was deleted, and the egress service was not configured, nothing to do.
+			return nil
 		}
-		// The service was deleted and was an egress service.
-		// We delete all of its relevant resources to avoid leaving stale configuration.
+		// The egress service is configured, but the service object was deleted,
+		// meaning we should clear all of its resources.
 		return c.clearServiceResourcesAndRequeue(key, state)
 	}
 
-	// At this point both the EgressService resource and the Service != nil
+	// At this point both the EgressService is assigned and the Service != nil
 
 	if state != nil && state.stale {
 		// The service is marked stale because something failed when trying to delete it.
 		// We try to delete it again before doing anything else.
+		klog.Warningf("Cleaning up stale egress service %s", key)
```

Comment:
I tried to stick to the approach from the original approach.
I think it would be better to treat this as an improvement and I can create an upstream issue for it, wdyt?

Reply:
yeah i added it as a topic to be discussed in our internal meeting tmrw.